rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+  ZNSession * s = node->context->impl->session;
+  service_data->zn_session_ = s;
+
+  // Obtain qualified request-response topics
+  std::string zn_request_topic_key(service->service_name);
+  zn_request_topic_key += "/request";
+
+  service_data->zn_request_topic_key_ = rcutils_strdup(zn_request_topic_key.c_str(),
+                                                       *allocator);
+
+  std::string zn_response_topic_key(service->service_name);
+  zn_response_topic_key += "/response";
+
+  service_data->zn_response_topic_key_ = rcutils_strdup(zn_response_topic_key.c_str(),
+                                                        *allocator);
+
+  // NOTE(CH3): This topic ID only unique WITHIN this process!
+  //
+  // Another topic on another process might clash with the ID on this process, even within the
+  // same Zenoh network! It is not a UUID!!
+  service_data->zn_response_topic_id_ = zn_declare_resource(s,
+                                                            service_data->zn_response_topic_key_);
+
+  // Init type support callbacks
+  auto service_members = static_cast<const service_type_support_callbacks_t *>(type_support->data);
+  auto request_members = static_cast<const message_type_support_callbacks_t *>(
+    service_members->request_members_->data
+  );
+  auto response_members = static_cast<const message_type_support_callbacks_t *>(
+    service_members->response_members_->data
+  );
+
+  service_data->typesupport_identifier_ = type_support->typesupport_identifier;
+  service_data->request_type_support_impl_ = request_members;
+  service_data->response_type_support_impl_ = response_members;
+
+  // Allocate and in-place assign new typesupport instances
+  service_data->request_type_support_ = static_cast<RequestTypeSupport_cpp *>(
+    allocator->allocate(sizeof(RequestTypeSupport_cpp *), allocator->state)
+  );
+  new(service_data->request_type_support_) RequestTypeSupport_cpp(service_members);
+  if (!service_data->request_type_support_) {
+    RMW_SET_ERROR_MSG("failed to allocate RequestTypeSupport");
+    allocator->deallocate(service_data->request_type_support_, allocator->state);
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
+    return nullptr;
+  }
+
+  service_data->response_type_support_ = static_cast<ResponseTypeSupport_cpp *>(
+    allocator->allocate(sizeof(ResponseTypeSupport_cpp *), allocator->state)
+  );
+  new(service_data->response_type_support_) ResponseTypeSupport_cpp(service_members);
+  if (!service_data->response_type_support_) {
+    RMW_SET_ERROR_MSG("failed to allocate ResponseTypeSupport");
+    allocator->deallocate(service_data->request_type_support_, allocator->state);
+    allocator->deallocate(service_data->response_type_support_, allocator->state);
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
```

Comment:
Does this leak `service->service_name`?

Reply:
Yep! :(

Updated: 23497d4430f890dc62fb9995c66fb2be8602b51e

(Sorry for the massive commit, it's kind of intertwined...)